src/eth-query.ts

```diff
@@ -0,0 +1,656 @@
+import createRandomIdFactory from 'json-rpc-random-id';
+
+const createRandomId = createRandomIdFactory();
+
+/**
+ * Makes a subset of an object type optional.
+ */
+type PickPartial<Type, Keys extends keyof Type> = { [Key in Keys]?: Type[Key] };
+
+/**
+ * What it says on the tin. We omit `null` as that value is used for a
+ * successful response to indicate a lack of an error.
+ */
+type EverythingButNull =
+  | string
+  | number
+  | boolean
+  // We mean to use `object` here.
+  // eslint-disable-next-line @typescript-eslint/ban-types
+  | object
+  | symbol
+  | undefined;
+
+/**
+ * A singular (non-collection) value within a request parameter list.
+ */
+type SimpleParam = string | number | boolean | null;
+
+/**
+ * The base type for the params that a request takes.
+ */
+type BaseParams = (SimpleParam | Record<string, SimpleParam>)[];
+
+/**
+ * The type of the `payload` for `sendAsync`.
+ */
+type SendAsyncPayload<Params extends BaseParams> = {
+  id: number;
+  jsonrpc: '2.0';
+  method: string;
+  params: Params;
+};
+
+/**
+ * The type of the `callback` for `sendAsync`.
+ */
+type SendAsyncCallback<Result> = (
+  ...args:
+    | [error: EverythingButNull, result: undefined]
+    | [error: null, result: Result]
+) => void;
+
+/**
+ * The type of the `response` for the provider object.
+ */
+type ProviderSendAsyncResponse<Result> = {
+  error?: { message: string };
+  result?: Result;
+};
+
+/**
+ * The type of the `callback` for the provider object's `sendAsync` method.
+ */
+type ProviderSendAsyncCallback<Result> = (
+  error: unknown,
+  response: ProviderSendAsyncResponse<Result>,
+) => void;
+
+/**
+ * An Ethereum provider object. Although it may contain other methods, all we
+ * care about in this library is `sendAsync`.
+ */
+export type Provider = {
+  sendAsync<Params extends BaseParams, Result>(
+    payload: SendAsyncPayload<Params>,
+    callback: ProviderSendAsyncCallback<Result>,
+  ): void;
+};
+
+/**
+ * Generates an instance method designed to call an RPC method. This instance
+ * method uses `sendAsync` internally to make the request to the network.
+ *
+ * @param methodName - The RPC method.
+ * @returns The generated method.
+ */
+function generateFnFor<Params extends BaseParams>(methodName: string) {
+  return function (
+    this: EthQuery,
+    ...args: [...Params, SendAsyncCallback<Params>]
+  ) {
+    const callback = args.pop();
+    // Typecast: The remaining arguments must be the params.
+    const params = args.slice(0, -1) as Params;
+    if (callback === undefined) {
```

Comment:
These checks are new. It makes sense that `callback` could be undefined because `args` may be empty. However, even though we've assigned a type of `[...Params, SendAsyncCallback<Params>]`, TypeScript isn't willing to assume that `callback` is `SendAsyncCallback<Params>`, but gives it a type of `SendAsyncCallback<Params> | Params[number]`.